File: src/python/pants/backend/go/util_rules/asm_check.py
# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
# Licensed under the Apache License, Version 2.0 (see LICENSE).

import sys

# Script that checks whether the files passed on the command line looks like they could be
# Golang-format assembly language files.
#
# This is used by the cgo rules as a heuristic to determine if the user is passing Golang assembly
# format instead of gcc assembly format.


def maybe_is_golang_assembly(data: bytes) -> bool:
    return (
        data.startswith(b"TEXT")
        or b"\nTEXT" in data
        or data.startswith(b"DATA")
        or b"\nDATA" in data
        or data.startswith(b"GLOBL")
        or b"\nGLOBL" in data
    )


def main(args):
    for arg in args:
        with open(arg, "rb") as f:
            data = f.read()
        if maybe_is_golang_assembly(data):
            print(f"{arg}")


if __name__ == "__main__":
    main(sys.argv[1:])

File: src/python/pants/backend/go/go_sources/asm_check/asm_check.go
/* Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
 * Licensed under the Apache License, Version 2.0 (see LICENSE).
 */

package main

import (
	"bytes"
	"fmt"
	"os"
)

func maybeGolangAssembly(filename string) (bool, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return false, err
	}

	if bytes.HasPrefix(data, []byte("TEXT")) || bytes.Contains(data, []byte("\nTEXT")) ||
		bytes.HasPrefix(data, []byte("DATA")) || bytes.Contains(data, []byte("\nDATA")) ||
		bytes.HasPrefix(data, []byte("GLOBL")) || bytes.Contains(data, []byte("\nGLOBL")) {
		return true, nil
	}

	return false, nil
}

func main() {
	if len(os.Args) >= 2 {
		for _, arg := range os.Args[1:] {
			found, err := maybeGolangAssembly(arg)
			if err != nil {
				fmt.Fprintf(os.Stderr, "ERROR: %v\n", err)
				os.Exit(1)
			}

			if found {
				fmt.Printf("%s\n", arg)
			}
		}
	}
}
